cmd/boulder-wfe2/main.go

```diff
@@ -57,6 +64,104 @@ type config struct {
+// validates that the PEM is well-formed with no leftover bytes, and contains
+// only a well-formed X509 certificate. If the chain file meets these
+// requirements the PEM bytes from the file are returned, otherwise an error is
+// returned.
+func loadChainFile(aiaIssuerURL, chainFile string) ([]byte, error) {
+	pemBytes, err := ioutil.ReadFile(chainFile)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"CertificateChain entry for AIA issuer url %q has an "+
+				"invalid chain file: %q - error reading contents: %s",
+			aiaIssuerURL, chainFile, err)
+	}
+	// Try to decode the contents as PEM
+	certBlock, rest := pem.Decode(pemBytes)
+	if certBlock == nil {
+		return nil, fmt.Errorf(
+			"CertificateChain entry for AIA issuer url %q has an "+
+				"invalid chain file: %q - contents did not decode as PEM",
+			aiaIssuerURL, chainFile)
+	}
+	// The PEM contents must be a CERTIFICATE
+	if certBlock.Type != "CERTIFICATE" {
+		return nil, fmt.Errorf(
+			"CertificateChain entry for AIA issuer url %q has an "+
+				"invalid chain file: %q - PEM block type incorrect, found "+
+				"%q, expected \"CERTIFICATE\"",
+			aiaIssuerURL, chainFile, certBlock.Type)
+	}
+	// The PEM Certificate must successfully parse
+	if _, err := x509.ParseCertificate(certBlock.Bytes); err != nil {
+		return nil, fmt.Errorf(
+			"CertificateChain entry for AIA issuer url %q has an "+
+				"invalid chain file: %q - certificate bytes failed to parse: %s",
+			aiaIssuerURL, chainFile, err)
+	}
+	// If there are bytes leftover we must reject the file otherwise these
+	// leftover bytes will end up in a served certificate chain.
+	if len(rest) != 0 {
+		return nil, fmt.Errorf(
+			"CertificateChain entry for AIA issuer url %q has an "+
+				"invalid chain file: %q - PEM contents had unused remainder "+
+				"input (%d bytes)",
+			aiaIssuerURL, chainFile, len(rest))
+	}
+
+	return pemBytes, nil
+}
+
+// loadCertificateChains processes the provided chainConfig of AIA Issuer URLs
+// and cert filenames. For each AIA issuer URL all of its cert filenames are
+// read, validated as PEM certificates, and concatenated together separated by
+// newlines. The combined PEM certificate chain contents for each are returned
+// in the results map, keyed by the AIA Issuer URL.
+func loadCertificateChains(chainConfig map[string][]string) (map[string][]byte, error) {
+	results := make(map[string][]byte, len(chainConfig))
+
+	// For each AIA Issuer URL we need to read the chain files
+	for aiaIssuerURL, chainFiles := range chainConfig {
+		var buffer bytes.Buffer
+
+		// There must be at least one chain file specified
+		if len(chainFiles) == 0 {
+			return nil, fmt.Errorf(
+				"CertificateChain entry for AIA issuer url %q has no chain "+
+					"file names configured",
+				aiaIssuerURL)
+		}
+
+		// Prepend a newline before the chain
+		buffer.Write([]byte("\n"))
+
+		// Read each chain file in reverse order - each is expected to be a PEM
```

Comment:
Why read these in reverse order? My expectation would be that if the config file lists `a.pem` and `b.pem`, the returned chain would have `[ee, a.pem, b.pem]`. Either way, this ambiguity points out that we should have a comment about ordering on the config option.

Reply:
> Either way, this ambiguity points out that we should have a comment about ordering on the config option.

A comment here? On the WFE `certificateChain` field? Both? Ideally it would be in the config file but we can't comment the JSON.

Reply:
Our approach so far has been to comment on the config structure in the main.go files. We still need to find a good way to make that into easily-readable documentation, but for now I think that's where we should collect it.